### ISSUE TITLE
Any advice with large embedding layer with large batch size?

### ISSUE THREAD

**Issue by TomorrowIsAnOtherDay** [uber/horovod#228]
Thanks for your nice job, Horovod can be used to train lots of  image models which have a lot of dense parameter.
But I found its performance with large embedding layer with large batch size is very slow,
I supposed its bottleneck is 
https://github.com/uber/horovod/blob/85882b0b04afe207ae9d5b91cc1eef726dea0083/horovod/tensorflow/__init__.py#L68
Because it have to collect  lots of sparse gradients and broadcast it to all workers.
Any advice on this task?

**Comment by alsrgv** [comment]
@TomorrowIsAnOtherDay, it would be great if we make Horovod work efficiently with sparse models as well.  Is there a good public benchmark which would approximate your workload so we could profile it internally?

**Comment by TomorrowIsAnOtherDay** [comment]
How about word2vec with wikipedia corpus?

**Comment by alsrgv** [comment]
@TomorrowIsAnOtherDay, how large is your sparse embedding dimensionality?

**Comment by TomorrowIsAnOtherDay** [comment]
Two embedding layers:
[100001, 256] 
[400001, 256]
And I tried to launch two mpi workers with batch size 4096 each.
total batch size is 4096 * 2.

**Comment by TomorrowIsAnOtherDay** [comment]
@alsrgv But [400001, 256] is an embedding layer used for word embedding.
In my case, a sentence(average length 100) was fed into memory for each example.

**Comment by alsrgv** [comment]
@TomorrowIsAnOtherDay, 400001 is vocabulary size, right? So, for each word you'll have 256 element vector. If an average length is 100, you'll do ~1.6GB per batch.  It's going to be harder to scale as the number of workers grows though - but doable, since the data can be pinned to CPU RAM.  What is the time per batch with distributed training and on a single GPU?

**Comment by TomorrowIsAnOtherDay** [comment]
a single GPU: 55 batch /s
two GPU: 7.27 batch /s

**Comment by alsrgv** [comment]
Right, this kind of workload would be hard to scale this way - you're essentially doing no computation, it's all about communication and matrix lookups.  Probably you don't even need GPUs.  This may benefit from having a parameter server with parameter matrix sharded across multiple servers.

**Comment by TomorrowIsAnOtherDay** [comment]
Thanks for your advice.
But then the bottleneck would be in worker?
Because the worker must sync its own model for parameter server before running the next batch, though having multiple servers.

**Comment by alsrgv** [comment]
Right, basically you have a set of parameter servers, each time you process batch you need to request embedding values at given indices from parameter server, compute updates and send them back.  You may benefit from running asynchronous training since there's less chance of collision between workers compared to the dense training of image models, and so you won't have a problem of stale gradients.

**Comment by TomorrowIsAnOtherDay** [comment]
We still have a lot of layers above embedding layers(e.g encoding/decoding).
Is asynchronous training still a good choice?

**Comment by TomorrowIsAnOtherDay** [comment]
Thanks for your advice：）I will have a try.

**Comment by UditGupta10** [comment]
I am facing a similar scalability problem for large batch sizes. @TomorrowIsAnOtherDay What's your take on this specific problem of yours, parameter server or horovod for scalability for large batch sizes?
@alsrgv What is the best way as of now (with horovod) to scale sparse updates with large batch sizes?
Thanks..
